src/login/logind-session.c

```diff
@@ -179,6 +179,20 @@ int session_set_leader(Session *s, pid_t pid) {
         return 1;
 }
 
+int session_invalidate_leader(Session *s) {
+        assert(s);
+
+        if (pid_is_alive(s->leader))
+                return 0;
+
+        (void) hashmap_remove_value(s->manager->sessions_by_leader, PID_TO_PTR(s->leader), s);
+        s->leader = 0;
+
+        (void) session_save(s);
```

Comment:
Maybe this whole block should be guarded by `s->leader > 0`? Do we know that the leader is set if we get to here?

Reply:
Good point, we should probably do that.